pkg/container-utils/cri/cri.go

```diff
@@ -54,17 +59,46 @@ func NewCRIClient(name, socketPath string, timeout time.Duration) (CRIClient, er
 		return CRIClient{}, err
 	}
 
-	return CRIClient{
+	cc := CRIClient{
 		Name:        name,
 		SocketPath:  socketPath,
 		ConnTimeout: timeout,
 		conn:        conn,
-		client:      pb.NewRuntimeServiceClient(conn),
-	}, nil
+		client:      runtime.NewRuntimeServiceClient(conn),
+	}
+
+	// determine CRI API version to use
+	if _, err = cc.client.Version(context.Background(), &runtime.VersionRequest{}); err == nil {
+		log.Debugf("Using CRI v1 for %s CRI client", name)
+	} else if status.Code(err) == codes.Unimplemented {
+		log.Debugf("Using CRI v1alpha2 for %s CRI client", name)
+		cc.client = nil
+		cc.clientV1alpha2 = runtimeV1alpha2.NewRuntimeServiceClient(conn)
+	} else {
+		log.Warnf("Failed to determine CRI API version: %v", err)
+	}
+
+	return cc, nil
 }
 
-func listContainers(c *CRIClient, filter *pb.ContainerFilter) ([]*pb.Container, error) {
-	request := &pb.ListContainersRequest{}
+func listContainers(c *CRIClient, filter *runtime.ContainerFilter) ([]*runtime.Container, error) {
+	if c.useV1alpha2() {
+		request := &runtimeV1alpha2.ListContainersRequest{}
+		if filter != nil {
+			request.Filter = &runtimeV1alpha2.ContainerFilter{
+				Id: filter.Id,
+			}
```

Comment:
I see that both versions define this structure in the same way 
https://github.com/kubernetes/cri-api/blob/3cbaf0d76da3429637d3eeb6de47045f50ebdc8b/pkg/apis/runtime/v1alpha2/api.pb.go#L4913 vs https://github.com/kubernetes/cri-api/blob/3cbaf0d76da3429637d3eeb6de47045f50ebdc8b/pkg/apis/runtime/v1/api.pb.go#L4948

Is it possible to define a function like `fromV1alpha2ListContainerFilter` to perform the conversion keep all the fields? (I think it can use the same exact approach as in `fromV1alpha2ListContainersResponse`)

Reply:
Ah! Right. Thanks for spotting it. I see it also in [kubelet/convertion.go](https://github.com/kubernetes/kubernetes/blob/v1.25.9/pkg/kubelet/cri/remote/conversion.go#L108) as well, somehow I missed it. I updated the PR now